src/service/evm.rs

```diff
@@ -0,0 +1,542 @@
+		&task_manager,
+		frontier_backend.clone(),
+	)?;
+
+	let filter_pool: FilterPool = Arc::new(Mutex::new(BTreeMap::new()));
+	let fee_history_cache: FeeHistoryCache = Arc::new(Mutex::new(BTreeMap::new()));
+
+	let params = PartialComponents {
+		backend,
+		client,
+		import_queue,
+		keystore_container,
+		task_manager,
+		transaction_pool,
+		select_chain: (),
+		other: (
+			block_import,
+			telemetry,
+			telemetry_worker_handle,
+			frontier_backend,
+			filter_pool,
+			fee_history_cache,
+		),
+	};
+
+	Ok(params)
+}
+
+/// Start a node with the given parachain `Configuration` and relay chain `Configuration`.
+///
+/// This is the actual implementation that is abstract over the executor and the runtime api.
+#[allow(clippy::too_many_arguments)]
+#[sc_tracing::logging::prefix_logs_with("Parachain")]
+pub(crate) async fn start_node_impl<RuntimeApi, Executor, RB, BIQ, BIC>(
+	parachain_config: Configuration,
+	polkadot_config: Configuration,
+	collator_options: CollatorOptions,
+	id: ParaId,
+	rpc_ext_builder: RB,
+	build_import_queue: BIQ,
+	build_consensus: BIC,
+) -> sc_service::error::Result<(TaskManager, Arc<FullClient<RuntimeApi>>)>
+where
+	RuntimeApi: ConstructRuntimeApi<Block, FullClient<RuntimeApi>> + Send + Sync + 'static,
+	RuntimeApi::RuntimeApi: sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+		+ sp_api::Metadata<Block>
+		+ sp_session::SessionKeys<Block>
+		+ sp_api::ApiExt<Block, StateBackend = sc_client_api::StateBackendFor<FullBackend, Block>>
+		+ sp_offchain::OffchainWorkerApi<Block>
+		+ sp_block_builder::BlockBuilder<Block>
+		+ cumulus_primitives_core::CollectCollationInfo<Block>
+		+ EthereumRuntimeRPCApi<Block>
+		+ ConvertTransactionRuntimeApi<Block>,
+	sc_client_api::StateBackendFor<FullBackend, Block>: sp_api::StateBackend<BlakeTwo256>,
+	Executor: sc_executor::NativeExecutionDispatch + 'static,
+	RB: Fn(
+			Arc<FullClient<RuntimeApi>>,
+			Arc<sc_transaction_pool::FullPool<Block, FullClient<RuntimeApi>>>,
+			DenyUnsafe,
+			SubscriptionTaskExecutor,
+			Arc<NetworkService<Block, Hash>>,
+			Arc<FrontierBackend<Block>>,
+			FilterPool,
+			FeeHistoryCache,
+			Arc<OverrideHandle<Block>>,
+			Arc<EthBlockDataCacheTask<Block>>,
+		) -> Result<rpc::RpcExtension, sc_service::Error>
+		+ 'static,
+	BIQ: FnOnce(
+		Arc<FullClient<RuntimeApi>>,
+		ParachainBlockImport<RuntimeApi>,
+		&Configuration,
+		Option<TelemetryHandle>,
+		&TaskManager,
+		Arc<FrontierBackend<Block>>,
+	) -> Result<
+		sc_consensus::DefaultImportQueue<Block, FullClient<RuntimeApi>>,
+		sc_service::Error,
+	>,
+	BIC: FnOnce(
+		Arc<FullClient<RuntimeApi>>,
+		ParachainBlockImport<RuntimeApi>,
+		Option<&Registry>,
+		Option<TelemetryHandle>,
+		&TaskManager,
+		Arc<dyn RelayChainInterface>,
+		Arc<sc_transaction_pool::FullPool<Block, FullClient<RuntimeApi>>>,
+		Arc<NetworkService<Block, Hash>>,
+		SyncCryptoStorePtr,
+		bool,
+	) -> Result<Box<dyn ParachainConsensus<Block>>, sc_service::Error>,
```

Comment:
I think this is my record of seeing bound lines in a function 🏅

Reply:
but bounds are cool 😎

Reply:
Maybe, as an idea for this case, we could use a trait similar to `Config` is used for pallets, declaring all associated types with the required bounds and passing only that high-level trait.

Reply:
I would normally agree that moving it to a trait would be good. But since the service code is usually the hardest to fix during a Polkadot upgrade, I think it is worthwhile to have it match what is in the Substrate/Frontier examples projects. This way we can more easily look at the changes made to those examples and copy the appropriate changes to our code.

The more we diverge from those example codebases, the harder Polkadot upgrades will be

Reply:
Sure! I simply left that thought 👍🏻